[lldb] Handle an empty SBMemoryRegionInfo from scripted process

jasonmolenda opened the pull request:
A scripted process implementation might return an SBMemoryRegionInfo object in its implementation of `get_memory_region_containing_address` which will have an address 0 and size 0, without realizing the problems this can cause.  Several algorithms in lldb will try to iterate over the MemoryRegions of the process, starting at address 0 and expecting to iterate up to the highest vm address, stepping by the size of each region, so a 0-length region will result in an infinite loop.  Add a check to Process::GetMemoryRegionInfo that rejects a MemoryRegion which does not contain the requested address; a 0-length memory region will therefor always be rejected.

rdar://139678032

jasonmolenda committed:
[lldb] Handle an empty SBMemoryRegionInfo from scripted process

A scripted process implementation might return an SBMemoryRegionInfo
object in its implementation of `get_memory_region_containing_address`
which will have an address 0 and size 0, without realizing the
problems this can cause.  One problem happens when an expression
is run, and lldb needs to find part of the target's virtual address
space to store the result (actually stored in lldb's host memory),
it uses IRMemoryMap::FindSpace() to do that.  If the process supports
GetMemoryRegionInfo, FindSpace will step through the allocated
memory ranges looking for an unused one.  But if every request is
a memory region with address 0 + size 0, this loop will never
terminate.  Detect this kind of response from a scripted process
plugin and return an error, so callers iterating over the address
space can exit.

Added a test where I copied dummy_scripted_process.py from the
TestScriptedProcess.py existing test, and added a bad
`get_memory_region_containing_address` implementation.

rdar://139678032

jasonmolenda commented:
I am uncertain where I should detect an improperly formatted memory range.  It could be in `Process::GetMemoryRegionInfo`, which would catch a bad gdb remote serial protocol response as well.  Or it could be localized to IRMemoryMap::FindSpace if we only wanted to avoid an infinite loop in that one place -- but there are other parts of the code that try to iterate over the address space in the target with these objects.  This patch adjusts it in the most specific place that fixed the issue I was looking at, to start with.

medismailben commented on lldb/test/API/functionalities/scripted_process_empty_memory_region/dummy_scripted_process.py:
```diff
@@ -0,0 +1,137 @@
+from lldb.plugins.scripted_process import ScriptedThread
+
+
+class DummyStopHook:
+    def __init__(self, target, args):
+        self.target = target
+        self.args = args
+
+    def handle_stop(self, exe_ctx, stream):
+        print("My DummyStopHook triggered. Printing args: \n%s" % self.args)
+        sp = exe_ctx.process.GetScriptedImplementation()
+        sp.handled_stop = True
+
+
```
You shouldn't need this

medismailben commented on lldb/test/API/functionalities/scripted_process_empty_memory_region/dummy_scripted_process.py:
```diff
@@ -0,0 +1,137 @@
+        )
+
+
+def __lldb_init_module(debugger, dict):
+    # This is used when loading the script in an interactive debug session to
+    # automatically, register the stop-hook and launch the scripted process.
+    if not "SKIP_SCRIPTED_PROCESS_LAUNCH" in os.environ:
+        debugger.HandleCommand(
+            "target stop-hook add -k first -v 1 -k second -v 2 -P %s.%s"
+            % (__name__, DummyStopHook.__name__)
+        )
+        debugger.HandleCommand(
+            "process launch -C %s.%s" % (__name__, DummyScriptedProcess.__name__)
+        )
+    else:
+        print(
+            "Name of the class that will manage the scripted process: '%s.%s'"
+            % (__name__, DummyScriptedProcess.__name__)
+        )
+        print(
+            "Name of the class that will manage the stop-hook: '%s.%s'"
+            % (__name__, DummyStopHook.__name__)
+        )
```
Or this

medismailben reviewed (approved):
LGTM with comments.

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedProcess.cpp:
```diff
@@ -288,8 +288,15 @@ Status ScriptedProcess::DoGetMemoryRegionInfo(lldb::addr_t load_addr,
     region = *region_or_err;
+    if (region.GetRange().GetRangeBase() == 0 &&
+        (region.GetRange().GetByteSize() == 0 ||
+         region.GetRange().GetByteSize() == LLDB_INVALID_ADDRESS)) {
```
Why would the size be `LLDB_INVALID_ADDRESS`? Did you mean to check that the range base isn't `LLDB_INVALID_ADDRESS`? 

Is this the condition you want?

```
(region.GetRange().GetRangeBase() == 0 || region.GetRange().GetRangeBase() == LLDB_INVALID_ADDRESS) && region.GetRange().GetByteSize() == 0) 
```

jasonmolenda replied:
Last week I had another similar bug where a gdb remote stub was returning start addr 0 size UINT64_MAX as a region response, and I was trying to reject that in addition to addr 0 size 0.  I haven't seen someone send that same incorrect response from a scripted process, though, I should just check for 0/0 here.

jasonmolenda replied:
I'm not wedded to it.  I'm still debating moving both of these checks up in to Process::GetMemoryRegionInfo.  I haven't done it yet for fear of unintended consequences, but any memory region info response that claims 0 bytes or the entire address space is meaningless.

labath replied:
A single region that covers the entire address space is fairly boring, but it is the kind of thing I might return as an "I don't know" value (perhaps I'm in some embedded environment where every address really is readable?). What's the problem with that kind of result?

That said, I think this check could be made more robust. `(0, 0)` is the likeliest response, but doesn't this mean that we would still loop if the plugin returns `(0,1)` instead? It sounds to me that if we check that:
- the returned region is of non-zero size
- the returned region actually includes the address being queried (I guess this kinda subsumes the previous item)
then we can guarantee forward progress no matter what the plugin returns (?)

labath replied:
> We had a gdb stub returning {0, UINT64_MAX} the other week and it broke IRMemory::FindSpace() which will avoid any memory region with read/write/execute flags if qMemoryRegionInfo packets are supported. The stub claimed the entire address space, FindSpace() said it could not find any address range available, and all expressions broke.

Okay, but what's the alternative? Picking a piece of memory that may overlap with some existing data? It sounds to me like the stub gets exactly what it asked for.

> Yeah, a range of {0, 1} would result in algorithms like FindSpace() looping for a very long time, and be nearly as bad. But so far the two instances I've seen of people return bad ranges are {0,0} and {0,UINT64_MAX}.

True, but if we can change the expression to catch both, why not do it? What I'm suggesting is to change the expression into something like `if (GetRange().GetRangeBase() > addr || GetRange().GetRangeEnd() <= addr)`. The `(0,0) case is subsumed by that, but this also catches any other incorrect response.

jasonmolenda replied:
I see what you mean for checking the address is contained within the range, and rejecting it if not, this works too.

Right now if a stub returns {0, UINT64_MAX} and we don't have a memory allocation packet, IRMemory::FindSpace will find no available virtual address space and all expressions will fail.

My opposition to {0, UINT64_MAX} being accepted is that it's meaningless.  No processor actually has 64-bits of addressing, so telling me that all memory is addressable is obviously wrong.  It will break IRMemoryMap::FindSpace's algorithm, and this response tells us nothing more than if qMemoryRegionInfo was an unsupported packet.

labath replied:
I guess the reason I don't like that is that it just feels like a very special case. Like if I send two regions, each covering half the address space, then I'm sort of saying the same thing, but unlike the single region, this response will get accepted. It also doesn't handle the case of a 32 bit address space being full. The fact that this breaks all of expression evaluation might be sort of a good thing actually, as it means the problem is likely to be caught.

If you feel strongly about this, then I don't want to stand in your way, but I would rather not have this special case.

jasonmolenda replied:
No, your points make sense, thanks for sticking with this discussion.  I'll update the patch as you suggest.

jasonmolenda committed:
Remove the clearly unnecessary parts from
dummy_scripted_process.py.  Add a comment in
ScriptedProcess::DoGetMemoryRegionInfo about
what types of memory regions are rejecting.

jasonmolenda committed:
Update patch to move the check for memory regions to
Process::GetMemoryRegionInfo - I want to reject these
regions from either gdb remote serial protocol stubs,
or scripted processes.

jasonmolenda committed:
Update error message.

jasonmolenda committed:
Change the test to be "returned region must contain addr"

This will reject a 0-length memory region response, e.g.
an empty SBMemoryRegionInfo object.

jasonmolenda commented:
I landed a fix with Pavel's suggestion, and rewrote the commit msg to reflect what the patch now does.

JDevlieghere reviewed (approved):
LGTM with the formatting fixed.

medismailben commented on lldb/source/Target/Process.cpp:
```diff
@@ -6184,7 +6184,14 @@ Status Process::GetMemoryRegionInfo(lldb::addr_t load_addr,
-  return DoGetMemoryRegionInfo(load_addr, range_info);
+  Status error = DoGetMemoryRegionInfo(load_addr, range_info);
+  // Reject a region that does not contain the requested address.
+  if (error.Success() && (range_info.GetRange().GetRangeBase() < load_addr ||
+                          range_info.GetRange().GetRangeEnd() <= load_addr))
```
I believe in `range_info.GetRange().GetRangeBase() < load_addr` the sign should be inverted.

You could use directly the `Range::{Contains, ContainsEndInclusive}` instead ?

```suggestion
  if (error.Success() && !range_info.GetRange().ContainsEndInclusive(load_addr))
```

medismailben reviewed (changes requested):
@jasonmolenda I think there is a mistake in your range check.

jasonmolenda committed:
Ismail caught a logic error on my part, used a built-in method
instead.

labath reviewed (commented):
Thanks Jason. Looks good. (I think the arm failure is due to it being a 32-bit machine. Maybe it would be sufficient to use 32-bit values everywhere?)

jasonmolenda commented:
> Thanks Jason. Looks good. (I think the arm failure is due to it being a 32-bit machine. Maybe it would be sufficient to use 32-bit values everywhere?)

I think the fail on the arm bot was because I have a scripted process plugin and it imports examples/python/templates/scripted_process.py which only has register definitions for arm64 and x86_64.  I added a skip for non-matching Targets shortly after I landed the patch.
